src/sqlfluff/rules/L064.py

```diff
@@ -0,0 +1,259 @@
+                fixes=[
+                    LintFix.replace(
+                        context.segment,
+                        [
+                            CodeSegment(
+                                raw=fixed_string,
+                                name="quoted_literal",
+                                type="literal",
+                            )
+                        ],
+                    )
+                ],
+                description=(
+                    "Inconsistent use of preferred quote style '"
+                    f"{self._quotes_mapping[preferred_string_quotes]['common_name']}'. "
+                    f"Use {fixed_string} instead of {context.segment.raw}."
+                ),
+            )
+
+        return None
+
+    def _sub_twice(self, regex: regex.Pattern, replacement: str, original: str) -> str:
+        """Replace `regex` with `replacement` twice on `original`.
+
+        This is used by string normalization to perform replaces on overlapping matches.
+        """
+        return regex.sub(replacement, regex.sub(replacement, original))
+
+    def _normalize_preferred_string_quotes(
+        self, s: str, preferred_quote_char: str, alternate_quote_char: str
+    ) -> str:
+        """Prefer `preferred_quote_char` but only if it doesn't cause more escaping.
+
+        Adds or removes backslashes as appropriate.
+        """
+        value = s.lstrip(self._string_prefix_chars)
+
+        if value[:3] == preferred_quote_char * 3:
+            # In tripple quoted strings we are not replacing escaped quotes.
+            # So nothing left to do and we can exit.
+            return s
+        elif value[0] == preferred_quote_char:
+            # Quotes are alright already. But maybe we can remove some unnecessary
+            # escapes or reduce the number of escapes using alternate_quote_char ?
+            orig_quote = preferred_quote_char
+            new_quote = alternate_quote_char
+        elif value[:3] == alternate_quote_char * 3:
+            orig_quote = alternate_quote_char * 3
+            new_quote = preferred_quote_char * 3
+        elif value[0] == alternate_quote_char:
+            orig_quote = alternate_quote_char
+            new_quote = preferred_quote_char
+        else:
+            self.logger.debug(
+                "Found quoted string %s using neither preferred quote char %s "
+                "nor alternate_quote_char %s. Skipping...",
+                s,
+                preferred_quote_char,
+                alternate_quote_char,
+            )
+            return s
+
+        first_quote_pos = s.find(orig_quote)
+        prefix = s[:first_quote_pos]
+        unescaped_new_quote = regex.compile(rf"(([^\\]|^)(\\\\)*){new_quote}")
+        escaped_new_quote = regex.compile(rf"([^\\]|^)\\((?:\\\\)*){new_quote}")
+        escaped_orig_quote = regex.compile(rf"([^\\]|^)\\((?:\\\\)*){orig_quote}")
+        body = s[first_quote_pos + len(orig_quote) : -len(orig_quote)]
```

Comment:
Just a thought -- I wonder if it'd be worthwhile to move this logic to a new property or function on the segment class. (For reuse by other rules, etc.)